File: examples/timer/timer.c
// From: http://docs.enlightenment.org/auto/ecore/Ecore_Main_Loop_Page.html

#include <Ecore.h>

Ecore_Timer         *timer1     = NULL;
Ecore_Event_Handler *handler1   = NULL;
double               start_time = 0.0;

int timer_func(void *data)
{
  printf("Tick timer. Sec: %3.2f\n", ecore_time_get() - start_time);
  return 1;
}

int exit_func(void *data, int ev_type, void *ev)
{
  Ecore_Event_Signal_Exit *e;

  e = (Ecore_Event_Signal_Exit *)ev;
  if (e->interrupt)      printf("Exit: interrupt\n");
  else if (e->quit)      printf("Exit: quit\n");
  else if (e->terminate) printf("Exit: terminate\n");
  ecore_main_loop_quit();
  return 1;
}

int main(int argc, const char **argv)
{
  ecore_init();
  ecore_app_args_set(argc, argv);  
  start_time = ecore_time_get();
  handler1 = ecore_event_handler_add(ECORE_EVENT_SIGNAL_EXIT, exit_func, NULL);
  timer1 = ecore_timer_add(0.5, timer_func, NULL);  
  ecore_main_loop_begin();
  ecore_shutdown();
  return 0;
}

File: examples/timer/timer.rb
#!/usr/bin/ruby

require '../ecore'

TimerCallback = Proc.new do
  puts "Tick timer. Sec: #{'%3.2f' % (Ecore::ecore_time_get - START_TIME)}"
  1
end

ExitCallback = Proc.new do |data, type, ev|
  event = Ecore::EcoreEventSignalExit.new(ev)
  if event[:interrupt]
    puts "Exit: interrupt"
  elsif event[:quit]
    puts "Exit: quit"
  elsif event[:terminate]
    puts "Exit: terminate"
  end
  Ecore::ecore_main_loop_quit()
  1
end

def string_pointers_from_array(array_of_strings)
  strptrs = array_of_strings.collect do |string|
    FFI::MemoryPointer.from_string(string)
  end
  strptrs << nil

  # Now load all the pointers into a native memory block
  array = FFI::MemoryPointer.new(:pointer, strptrs.length)
  strptrs.each_with_index do |p, i|
   array[i].put_pointer(0,  p)
  end
  array
end

Ecore::ecore_init
Ecore::ecore_app_args_set(ARGV.length, string_pointers_from_array(ARGV))
START_TIME = Ecore::ecore_time_get
handler1 = Ecore::ecore_event_handler_add(Ecore::ECORE_EVENT_SIGNAL_EXIT, ExitCallback, nil)
timer1 = Ecore::ecore_timer_add(0.5, TimerCallback, nil)
Ecore::ecore_main_loop_begin
Ecore::ecore_shutdown
